File: src/main/java/com/icube/sim/tichu/games/tichu/mappers/TrickMapper.java
package com.icube.sim.tichu.games.tichu.mappers;

import com.icube.sim.tichu.games.tichu.dtos.TrickDto;
import com.icube.sim.tichu.games.tichu.tricks.*;

import java.util.List;

public class TrickMapper {
    private final CardMapper cardMapper = new CardMapper();

    public TrickDto toDto(Trick trick) {
        var builder = TrickDto.builder()
                .playerIndex(trick.getPlayerIndex())
                .type(trick.getType())
                .cards(cardMapper.toDtos(trick.getCards()));

        switch (trick) {
            case SingleTrick singleTrick -> {
                var rank = singleTrick.getRank();
                builder.rank(rank).phoenixRank(singleTrick.isPhoenixUsed() ? rank : null);
            }
            case PairTrick pairTrick -> {
                var rank = (float) pairTrick.getRank();
                builder.rank(rank).phoenixRank(pairTrick.isPhoenixUsed() ? rank : null);
            }
            case ThreeOfAKindTrick threeOfAKindTrick -> {
                var rank = (float) threeOfAKindTrick.getRank();
                builder.rank(rank).phoenixRank(threeOfAKindTrick.isPhoenixUsed() ? rank : null);
            }
            case FullHouseTrick fullHouseTrick -> {
                var rank = (float) fullHouseTrick.getRank();
                var phoenixRank = fullHouseTrick.getPhoenixRank();
                builder.rank(rank).phoenixRank(phoenixRank == null ? null : Float.valueOf(phoenixRank));
            }
            case ConsecutivePairsTrick consecutivePairsTrick -> {
                var phoenixRank = consecutivePairsTrick.getPhoenixRank();
                builder.minRank(consecutivePairsTrick.getMinRank())
                        .maxRank(consecutivePairsTrick.getMaxRank())
                        .phoenixRank(phoenixRank == null ? null : Float.valueOf(phoenixRank));
            }
            case StraightTrick straightTrick -> {
                var phoenixRank = straightTrick.getPhoenixRank();
                builder.minRank(straightTrick.getMinRank())
                        .maxRank(straightTrick.getMaxRank())
                        .phoenixRank(phoenixRank == null ? null : Float.valueOf(phoenixRank));
            }
            case DogTrick ignored -> {}
            case FourOfAKindTrick fourOfAKindTrick -> {
                builder.rank((float) fourOfAKindTrick.getRank());
            }
            case StraightFlushTrick straightFlushTrick -> {
                builder.minRank(straightFlushTrick.getMinRank()).maxRank(straightFlushTrick.getMaxRank());
            }
            case null, default -> throw new IllegalArgumentException("Invalid trick type");
        }

        return builder.build();
    }

    public List<TrickDto> toDtos(List<Trick> tricks) {
        return tricks.stream().map(this::toDto).toList();
    }
}

File: frontend/src/games/tichu/mappers/TrickMapper.ts
import { TrickDto } from "@/games/tichu/dtos/TrickDto.ts";
import {
  ConsecutivePairsTrick, DogTrick, FourOfAKindTrick,
  FullHouseTrick,
  PairTrick,
  SingleTrick, StraightFlushTrick, StraightTrick,
  ThreeOfAKindTrick,
  Trick,
  TrickType
} from "@/games/tichu/domain/Trick.ts";
import { CardMapper } from "@/games/tichu/mappers/CardMapper.ts";
import { CardRank, CardSuit, StandardCard } from "@/games/tichu/domain/Card.ts";

export class TrickMapper {
  static toTrick(trickDto: TrickDto): Trick {
    const cards = trickDto.cards.map(CardMapper.toCard);
    switch (trickDto.type) {
      case TrickType.SINGLE:
        if (trickDto.phoenixRank !== null) {
          const prevRank = trickDto.phoenixRank - 0.5 as CardRank;
          const dummyCard = new StandardCard(CardSuit.SPADE, prevRank);
          const dummyTrick = new SingleTrick(0, [dummyCard], null);
          return new SingleTrick(trickDto.playerIndex, cards, dummyTrick);
        } else {
          return new SingleTrick(trickDto.playerIndex, cards, null);
        }
      case TrickType.PAIR:
        return new PairTrick(trickDto.playerIndex, cards);
      case TrickType.THREE_OF_A_KIND:
        return new ThreeOfAKindTrick(trickDto.playerIndex, cards);
      case TrickType.FULL_HOUSE:
        return new FullHouseTrick(trickDto.playerIndex, cards);
      case TrickType.CONSECUTIVE_PAIRS:
        return new ConsecutivePairsTrick(trickDto.playerIndex, cards);
      case TrickType.STRAIGHT:
        return new StraightTrick(trickDto.playerIndex, cards);
      case TrickType.DOG:
        return new DogTrick(trickDto.playerIndex, cards);
      case TrickType.FOUR_OF_A_KIND:
        return new FourOfAKindTrick(trickDto.playerIndex, cards);
      case TrickType.STRAIGHT_FLUSH:
        return new StraightFlushTrick(trickDto.playerIndex, cards);
    }
  }
}
